ARROW-12168: [Go][IPC] Implement Compression handling for Arrow IPC

zeroshade opened the pull request:
This is my follow up from #9836

emkornfield commented:
I'll try to review soon if @sbinet  can't but could you please remove the block on testing Go's ability to read integration test files and update docs for go in this PR (https://github.com/apache/arrow/pull/9822 also changed relevant files).  Note integration tests in CI are currently broken so you will have to run archery by hand to verify Go can read the files.

zeroshade commented:
Ran archery locally with `--with-go true` and it successfully worked with the compression test cases.

emkornfield commented on go/arrow/go.sum:
```diff
@@ -55,13 +67,11 @@ golang.org/x/sync v0.0.0-20180314180146-1d60e4601c6f/go.mod h1:RxMgew5VJxzue5/jJ
 golang.org/x/sys v0.0.0-20200323222414-85ca7c5b95cd/go.mod h1:h1NjWce9XRLGQEsW7wpKNCjG9DtNlClVuFLEZdDNbEs=
 golang.org/x/sys v0.0.0-20200909081042-eff7692f9009 h1:W0lCpv29Hv0UaM1LXb9QlBHLNP8UFfcKjblhVCWftOM=
 golang.org/x/sys v0.0.0-20200909081042-eff7692f9009/go.mod h1:h1NjWce9XRLGQEsW7wpKNCjG9DtNlClVuFLEZdDNbEs=
-golang.org/x/text v0.3.0 h1:g61tztE5qeGQ89tm6NTjjM9VPIm088od1l6aSorWRWg=
```
it looks this is getting swapped for kr/text?  why the change?

zeroshade replied:
It's not a change i made directly so my guess is that something underneath changed a dependency and I picked it up.

emkornfield commented on go/arrow/internal/arrdata/ioutil.go:
```diff
@@ -172,6 +173,47 @@ func WriteFile(t *testing.T, f *os.File, mem memory.Allocator, schema *arrow.Sch
 }
 
+// WriteFile writes a list of records to the given file descriptor, as an ARROW file.
+func WriteFileCompressed(t *testing.T, f *os.File, mem memory.Allocator, schema *arrow.Schema, recs []array.Record, codec flatbuf.CompressionType) {
```
at least in other languages I'm familiar with there is some effort to hide the flat buffer details from consumers of the API.  Would it make sense to abstract that here?

zeroshade replied:
So this is an internal package (note the "internal" in the path) which means it cannot be imported or called outside of the `arrow` repo here. This is only used for the unit tests so I figured it wasn't a problem to have it use the enum from the flatbuffer details directly in this situation since it's not an externally exposed API for consumers.

The only externally exposed API for consumers is the `WithLZ4` and `WithZstd` functions that were added as options to be passed to the writer which explicitly doesn't expose the flatbuffer details to external consumers.

emkornfield commented on go/arrow/internal/arrdata/ioutil.go:
```diff
@@ -172,6 +173,47 @@ func WriteFile(t *testing.T, f *os.File, mem memory.Allocator, schema *arrow.Sch
+		opts = append(opts, ipc.WithLZ4())
+	case flatbuf.CompressionTypeZSTD:
+		opts = append(opts, ipc.WithZstd())
+	}
```
fail or at least warn on behavior for unrecognized type?

zeroshade replied:
sure i'll add a failure here for that.

emkornfield replied:
Seems like errors in this method are returned via `t`, Should the same be done there?

zeroshade replied:
ah fair point, this is a testing function I should just use the `t`. I'll fix that

zeroshade replied:
fixed

emkornfield commented on go/arrow/internal/arrdata/ioutil.go:
```diff
@@ -172,6 +173,47 @@ func WriteFile(t *testing.T, f *os.File, mem memory.Allocator, schema *arrow.Sch
+		t.Fatalf("could not sync data to disk: %v", err)
+	}
+
+	_, err = f.Seek(0, io.SeekStart)
```
why the seek to start?

zeroshade replied:
So, I actually copied the existing WriteFile test function and then just added the compression to it, but looking at the code my guess is that it's because the file isn't being opened here, it's passed in. So doing the seek to start after writing just preps the file handle to immediately be able to read from it for the test rather than leaving it at the end of the file after doing the writes, which makes sense since this is only used for testing and can't be called externally

emkornfield replied:
OK, maybe add a comment.

zeroshade replied:
sure thing

zeroshade replied:
comment added.

emkornfield commented on go/arrow/internal/arrdata/ioutil.go:
```diff
@@ -191,3 +233,32 @@ func WriteStream(t *testing.T, f *os.File, mem memory.Allocator, schema *arrow.S
+
+// WriteStreamCompressed writes a list of records to the given file descriptor as an ARROW stream
+// using the provided compression type.
+func WriteStreamCompressed(t *testing.T, f *os.File, mem memory.Allocator, schema *arrow.Schema, recs []array.Record, codec flatbuf.CompressionType) {
```
same comment on the api.

zeroshade replied:
same as i mentioned above, this is not an externally facing Consumable API because it's an "internal" package which cannot be imported outside of this repo. The "internal" package is a special package that prevents anything below it from being an externally facing public API.

emkornfield commented on go/arrow/ipc/compression.go:
```diff
@@ -0,0 +1,106 @@
+
+// from zstd.h
+func (zstdCompressor) MaxCompressedLen(len int) int {
+	extra := ((128 << 10) - len) >> 11
```
please be careful here with signed values here.   it looks like ZSTD_compressBound takes size_t which I believe is unsigned.

zeroshade replied:
updated to use `uint` here to ensure that we don't make any mistakes with signed vs unsigned values.

emkornfield commented on go/arrow/ipc/compression.go:
```diff
@@ -0,0 +1,106 @@
+	*zstd.Encoder
+}
+
+// from zstd.h
```
mention the method name/macro here where it was taken from.

emkornfield replied:
It is strange this isn't provided as part of the library?

zeroshade replied:
yea it's very strange and slightly annoying.

zeroshade replied:
Added relevant comments.

emkornfield commented on go/arrow/ipc/compression.go:
```diff
@@ -0,0 +1,106 @@
+	switch codec {
+	case flatbuf.CompressionTypeLZ4_FRAME:
+		w := lz4.NewWriter(nil)
+		w.Apply(lz4.ConcurrencyOption(1), lz4.ChecksumOption(false), lz4.BlockSizeOption(lz4.Block64Kb))
```
maybe comment on how these options are chosen?

zeroshade replied:
added comment about the options, which were chosen to match the defaults that the C++ implementation uses

emkornfield commented on go/arrow/ipc/compression.go:
```diff
@@ -0,0 +1,106 @@
+	case flatbuf.CompressionTypeZSTD:
+		enc, err := zstd.NewWriter(nil)
+		if err != nil {
+			panic(err)
```
shouldn't errors be propogated up?

zeroshade replied:
as mentioned in my other comment, i was following the current existing pattern for the arrow library in using panic for these situations. In theory this shouldn't happen unless something has gone seriously wrong and if so, it's not recoverable except to try again without the compression. The current pattern in the arrow library seems to be to panic in these situations so that's what I maintained

emkornfield commented on go/arrow/ipc/compression.go:
```diff
@@ -0,0 +1,106 @@
+	case flatbuf.CompressionTypeZSTD:
+		dec, err := zstd.NewReader(nil)
+		if err != nil {
+			panic(err)
```
same question on error management.

emkornfield commented on go/arrow/ipc/file_reader.go:
```diff
@@ -312,16 +312,23 @@ func (f *FileReader) ReadAt(i int64) (array.Record, error) {
 	var (
-		msg = flatbuf.GetRootAsMessage(meta.Bytes(), 0)
-		md  flatbuf.RecordBatch
+		msg   = flatbuf.GetRootAsMessage(meta.Bytes(), 0)
```
what applied this formatting change? Is go formatting not already checked in CI?

zeroshade replied:
I don't know if go formatting is checked in the CI already, but the formatting change here happened because my IDE is set to automatically apply gofmt formatting whenever I save. So this formatting change happened because of applying proper formatting.

emkornfield commented on go/arrow/ipc/file_reader.go:
```diff
@@ -348,10 +356,26 @@ func (src *ipcSource) buffer(i int) *memory.Buffer {
+		raw = make([]byte, buf.Length())
+		_, err := src.r.ReadAt(raw, buf.Offset())
+		if err != nil {
+			panic(err)
```
seems like panic is used in existing code, this seems strange to me though.

zeroshade replied:
the choice to use panic is probably to make the consumable api easier to manage instead of having to propagate and push errors everywhere it just fails hard. Since the existing code was already using panic, that's what I maintained in my changes rather than adding error out params.

emkornfield replied:
seems OK to me.  I don't know Go well enough but for IO related code I would expect failures at least some of the time and I thought panic was meant for actual "stop the entire program" type errors.

zeroshade replied:
`panic` is relatively equivalent to throwing an exception in C++ code. You can recover from a panic just like you could catch a thrown exception in C++ code. The "general" advice between returning an error vs throwing a panic is whether or not the error happened during *normal* operation or happened due to something being coded/programmed incorrectly (like a pre/post-condition being violated etc.) 

As i mentioned, normally *I* would probably return the error here rather than panic, but I was maintaining the current pattern that existed in the code rather than creating a breaking change which would make consumers have to update their call-sites to handle the extra error out param.

emkornfield commented on go/arrow/ipc/file_reader.go:
```diff
@@ -348,10 +356,26 @@ func (src *ipcSource) buffer(i int) *memory.Buffer {
+	} else {
+		sr := io.NewSectionReader(src.r, buf.Offset(), buf.Length())
+		var uncompressedSize uint64
+		if err := binary.Read(sr, binary.LittleEndian, &uncompressedSize); err != nil {
```
I think this needs to check for the magic uncompressed value as well and revert back to a slice? Are zero length buffers handled upstream from this code?

zeroshade replied:
zero length buffers are handled above this in this function, if the buffer is length 0, this function returns a buffer with a nil slice. In addition the writer doesn't attempt to compress or write for a 0 length buffer either.

emkornfield reviewed (changes requested):
Sorry, realized I was leaving one off comments.  I finished my first pass, and I think the two biggest issues are:
1.  signed vs unsigned on ZSTD max compress size.
2.  checking for uncompressed buffers.

One other thing not necessary for this PR to consider (and it might not be possible with existing code structure) is decoupling compression/decompression and IO calls on separate threads.

emkornfield commented:
Also, I'm not sure if it has been merged yet but I saw a PR that appears to fix the integration test CI, so rebasing might actually get you a successful run.

zeroshade commented:
Rebased as per suggestion to hopefully get a successful run there.

As for the requested changes:

1) Updated the `MaxCompressLen` function for ZSTD to explicitly handle signed vs unsigned to ensure no issues there.
2) Is there a test case that exists in the integration archery tests which has the metadata say to use compression but passes uncompressed buffers? The code currently correctly handles 0 length and nil buffers, but I didn't see anything in the spec or config (or the C++ implementation) for having non-zero length buffers which are uncompressed being sent along side compressed ones. I guess I'm missing something?

As far as the decoupling of compression/decompression and IO calls:
For Decompression: The current code structure only decompresses as it loads the records but doesn't provide any decoupling between reading the record in with the IO calls and the loading of the flatbuffer data that I could put the decompression into. Currently it's just going to decompress in the same place that it would be loading the raw flatbuffer data into an `array.Record`.

For Compression: 
The compression happens during the payload encoding. So while the current code does not provide any hook that could be used to decouple encoding the payload and the IO to send it (the encoding happens right before it writes the record) if that decoupling were to be added, it is easy to do it as the compression is part of the payload encoding, not part of the IO to write the record out.

The one optimization I can see here is that I could implement parallelization for the compression to attempt to compress the body buffers in parallel rather than serially like I'm doing right now.

Thoughts?

emkornfield commented:
> Is there a test case that exists in the integration archery tests which has the metadata say to use compression but passes uncompressed buffers? The code currently correctly handles 0 length and nil buffers, but I didn't see anything in the spec or config (or the C++ implementation) for having non-zero length buffers which are uncompressed being sent along side compressed ones. I guess I'm missing something?

Spec reference is here https://github.com/apache/arrow/blob/master/format/Message.fbs#L59  Looking at the C++ code it doesn't look like it is conformance with the specification here.  I opened https://issues.apache.org/jira/browse/ARROW-12196

> The one optimization I can see here is that I could implement parallelization for the compression to attempt to compress the body buffers in parallel rather than serially like I'm doing right now.

In C++ at least threading is configurable.  I don't have a strong preference here, and really profiling for specific use-cases is probably important.  For small batches with narrow schemas I can imagine multithreading being slower then single threaded.   So feel free to leave as is if someone runs into performance issues using these features we ca explore the options.

zeroshade commented:
> Spec reference is here https://github.com/apache/arrow/blob/master/format/Message.fbs#L59 Looking at the C++ code it doesn't look like it is conformance with the specification here. I opened https://issues.apache.org/jira/browse/ARROW-12196

Okay, looks pretty simple, just check for an uncompressed length of -1 to indicate the data is not compressed. I'll update and add that.

> In C++ at least threading is configurable. I don't have a strong preference here, and really profiling for specific use-cases is probably important. For small batches with narrow schemas I can imagine multithreading being slower then single threaded. So feel free to leave as is if someone runs into performance issues using these features we ca explore the options.

So I know that the LZ4 library I'm using says it uses `runtime.GOMAXPROCS` to parallelize compression of a chunk of data, but it's fairly trivial to add the ability to parallelize the compression of the body buffers with an argument to control the parallelization. Thankfully goroutines are much lighter weight than full threads so it's less likely for there to be a performance degradation in the case of a small batch with narrow schema. I'll make the changes and if it turns out to be more complex than I'm thinking, then I'll agree with you that I'll leave it out until someone runs into a performance issue with the feature. If it's as simple as i'm thinking it'll be, then i'll just include it here.

zeroshade commented:
Newest version correctly checks for the -1 in the uncompressed length to handle uncompressed buffers correctly and implements the ability to specify a parallelization parameter for compression. Defaults to serial with no parallelization.

emkornfield commented on go/arrow/ipc/compression.go:
```diff
@@ -0,0 +1,107 @@
+	if len >= (128 << 10) {
+		extra = 0
+	}
+	return int(uint(len+(len>>8)) + extra)
```
might be worth an assertion `len >= 0` (I think I'm ok without this).

emkornfield commented on go/arrow/ipc/ipc.go:
```diff
@@ -104,6 +108,32 @@ func WithSchema(schema *arrow.Schema) Option {
 
+// WithLz4 tells the writer to use LZ4 Frame compression on the data
+// buffers before writing. Requires >= Arrow 1.0.0 to read/decompress
+func WithLZ4() Option {
```
small style nit:  Is there a go standard for LZ4 vs Lz4? (the documentation and method name are inconsistent here).

zeroshade replied:
The standard would be the way the function is named as `LZ4`, i'll fix the doc.

emkornfield reviewed (approved):
A few small nits, otherwise LGTM.

emkornfield commented:
Thanks.  I'll leave this open until at least Tuesday evening US/Pacific to see if @sbinet wants to add any comments.

zeroshade commented:
@emkornfield bump on getting this merged :)

emkornfield commented:
+1.  Thank you.
